db/snapshot/backup_task.cc

```diff
@@ -103,52 +105,131 @@ future<> backup_task_impl::upload_component(shared_ptr<s3::client> client, sstri
+                //   1. There are no more files.
+                //   2. An exception is thrown.
+                //   3. An abort is requested.
+                while (true) {
+                    // One of two things happened: the `upload_component` threw an exception, `aborter`'s exception was set.
+                    // The later can happen if the `aborter` was requested to abort the backup by task's abort source or another shard requested to abort since
+                    // it got an exception
+                    if (ex || aborter.abort_requested()) {
+                        break;
+                    }
+
+                    // Fetching files sequentially within each shard allows less loaded shards to take on more work, compensating for those that are lagging behind.
+                    auto name_idx = counter++;
+                    if (name_idx >= backup_files.size()) {
+                        break;
+                    }
+                    auto gh = uploads.hold();
+                    // Pre-upload break point. For testing abort in actual s3 client usage.
+                    co_await utils::get_local_injector().inject("backup_task_pre_upload", utils::wait_for_message(std::chrono::minutes(2)));
+
+                    // Start uploading the component asynchronously.
+                    // Future is ignored here because:
+                    // - The client (cln) is maintained via shared ownership.
+                    // - Filenames are passed by value.
+                    // - The abort source remains valid for the duration of the function.
+                    // - The gate ensures that execution remains until completion.
+                    std::ignore = upload_component(cln, aborter, _progress_per_shard[shard_id], std::get<1>(backup_files[name_idx]))
+                                      .handle_exception([&ex](std::exception_ptr e) {
+                                          // Record the exception to eventually abort the rest of the backup.
+                                          ex = std::move(e);
+                                      }).finally([gh = std::move(gh)] {});
+
+                    co_await coroutine::maybe_yield();
+                    co_await utils::get_local_injector().inject("backup_task_pause", utils::wait_for_message(std::chrono::minutes(2)));
+                }
+            } catch (...) {
+                // Catch any exception thrown during backup processing.
+                ex = std::current_exception();
+            }
+            co_await uploads.close();
+
+            // If the `aborter` was requested to abort the backup, no need to propagate the exception to other shards since it can be initiated by two ways:
+            // 1. The task's abort source requested to abort the backup on all shards (see the subscription creation).
+            // 2. Another shard requested to abort the backup since it got an exception.
+            if (ex && !aborter.abort_requested()) {
```

Comment:
This is racy. In case several shards get here with an `ex`, they go and try to do smp-abort of each other. So it's smp::count*nr_failer_shards complex. Chances are low, nonetheless

Reply:
it is, but calling abort on already aborted source is (almost) no-op, I think we can survive this storm since it is already unhappy path where we are aborting the work